packages/firestore/test/integration/api/server_timestamp.test.ts

```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
+
+  it('can return previous value of different type', () => {
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsUsePreviousValue(snapshot, null))
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          verifyTimestampsAreResolved(snapshot);
+        })
+        .then(() =>
+          docRef.update('a', firebase.firestore.FieldValue.serverTimestamp())
+        )
+        .then(waitForLocalEvent)
+        .then(snapshot => {
+          expect(snapshot.get('a', { serverTimestamps: 'none' })).to.equal(
+            null
+          );
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
+          expect(snapshot.get('a', { serverTimestamps: 'previous' })).to.equal(
+            42
+          );
+        })
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          expect(
+            snapshot.get('a', { serverTimestamps: 'none' })
+          ).to.be.instanceof(Date);
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
+          expect(
+            snapshot.get('a', { serverTimestamps: 'previous' })
+          ).to.be.instanceof(Date);
+        });
+    });
+  });
+
+  it('can return previous value through consecutive updates', () => {
+    return withTestSetup(() => {
+      // TODO(mikelehen): Find better way to expose this to tests.
+      // tslint:disable-next-line:no-any enableNetwork isn't exposed via d.ts
+      const firestoreInternal = docRef.firestore.INTERNAL as any;
+
+      let promises: Promise<void>[] = [];
+
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => {
+          verifyTimestampsUsePreviousValue(snapshot, null);
+          return waitForRemoteEvent();
+        })
+        .then(firestoreInternal.disableNetwork)
+        .then(() => {
+          promises.push(
+            docRef.update('a', firebase.firestore.FieldValue.serverTimestamp())
+          );
+          return waitForLocalEvent();
+        })
+        .then(snapshot => {
+          expect(snapshot.get('a', { serverTimestamps: 'previous' })).to.equal(
+            42
+          );
+          promises.push(
+            docRef.update('a', firebase.firestore.FieldValue.serverTimestamp())
+          );
+          return waitForLocalEvent();
+        })
+        .then(snapshot => {
+          expect(snapshot.get('a', { serverTimestamps: 'previous' })).to.equal(
+            42
+          );
+          return firestoreInternal.enableNetwork();
+        })
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          return Promise.all(promises);
+        })
+        .then(() => {});
+    });
+  });
+
+  it('uses previous value from local mutation', () => {
```

Comment:
It took me a couple minutes to figure out what was different about this test vs the previous one.  These tests really need less boilerplate and/or more comments. Or perhaps better would actually be to move these to unit tests rather than integration tests so you can more directly test the applyToLocalView() code without all of this extra setup / asynchrony / etc. in the way... :-/  The integration tests are for hitting all of the public surface area and sanity-checking behavior rather than covering all edge cases.

Reply:
I took a quick look at testing this via the mutations batches earlier, and some of this testing can indeed be moved there by verifying `previousState` via `valueWithOptions`.
Since Server Timestamps were already mostly tested via these integration tests, I ended up shying away from this though. 
Before I start reworking this, I would like to know how strong you feel about it, as this change should then also be ported to the Android and iOS clients.